source/sessions/snapshot-sessions.rst

```diff
@@ -0,0 +1,358 @@
+    both the primary and secondary nodes, including in sharded clusters.
+
+Snapshot cluster time
+    Snapshot cluster time, representing timestamp of the first read (find/aggregate operations) in the session.
+    The server creates a cursor in response to a snapshot read command and 
+    reports ``atClusterTime`` field in the cursor response. ``atClusterTime`` field represents the timestamp
+    of the read and is guaranteed to be majority committed.
+
+Specification
+=============
+
+An application requests snapshot reads by creating a ``ClientSession``
+with options that specify that snapshot reads are desired. An
+application then passes the session as an argument to methods in the
+``MongoDatabase`` and ``MongoCollection`` classes. Read operations (find/aggregate/distinct) performed against
+that session will be read from the same snapshot.
+
+High level summary of the API changes for snapshot reads
+========================================================
+
+Snapshot reads are built on top of client sessions.
+
+Applications will start a new client session for snapshot reads like
+this:
+
+.. code:: typescript
+
+    options = new SessionOptions(isSnapshot = true);
+    session = client.startSession(options);
+
+All read operations performed using this session will be read from same snapshot.
+
+If no value is provided for ``isSnapshot`` a value of false is
+implied.
+
+MongoClient changes
+===================
+
+There are no API changes to ``MongoClient`` to support snapshot reads.
+Applications indicate whether they want snapshot reads by setting the
+``isSnapshot`` field in the options passed to the ``startSession`` method.
+
+SessionOptions changes
+======================
+
+``SessionOptions`` change summary
+
+.. code:: typescript
+
+    class SessionOptions {
+        Optional<bool> isSnapshot;
+
+        // other options defined by other specs
+    }
+
+In order to support snapshot reads a new property named
+``isSnapshot`` is added to ``SessionOptions``. Applications set
+``isSnapshot`` when starting a client session to indicate
+whether they want snapshot reads. All read operations performed
+using that client session will share the same snapshot.
+
+Each new member is documented below.
+
+isSnapshot
+---------
+
+Applications set ``isSnapshot`` when starting a session to
+indicate whether they want snapshot reads.
+
+Note that the ``isSnapshot`` property is optional. The default value of
+this property is false.
+
+Snapshot reads and causal consistency are mutually exclusive. Therefore if ``isSnapshot`` is set to true,
+``causalConsistency`` property is set to false. Client MUST throw an Error if both ``isSnapshot`` and ``causalConsistency`` are set to true.
+Snapshot reads are supported both on primaries and secondaries.
+
+ClientSession changes
+=====================
+
+``ClientSession`` changes summary
+
+.. code:: typescript
+
+    interface ClientSession {
+        Optional<BsonTimestamp> snapshotClusterTime;
```

Comment:
"snapshotTimestamp"? I now forget whether we've discussed the user-facing terminology. It could be "snapshotClusterTime" to discourage users from providing their own timestamps (which might not work), but "snapshotTimestamp" seems more understandable and accurate to me.

Reply:
Done.